### ISSUE TITLE
optimize code representations

### ISSUE THREAD

**Issue by mndrix** [mndrix/golog#13]
Codes and lists of codes are both common Prolog data structures.  Create internal representations which optimize for this use case.  These representations will occupy less memory (the bottleneck in most benchmarks).

Make a `Code` type which implements the `Number` interface.  It should just be a `rune` under the hood.  The current implementation for integers requires a pointer to a `struct` which wraps a `bool` and a slice.  A `Code` type would also be helpful for representing small integers, which are very common in real world code.

Make a `CodeList` which implements the `Callable` interface.  It should just be a `[]rune` under the hood.  Unification with other list terms returns a slice where possible.  The current implementation for code lists requires a linked list of terms which adds substantial memory overhead.

**Comment by least-olegs** [comment]
Are you still working on this / have you done any progress?

I've written some helper functions for more compact printing of lists and strings. If you are interested, I could put it into a pull request.

**Comment by mndrix** [comment]
> Are you still working on this / have you made any progress?

No.  I haven't worked on this issue in quite a while and probably won't return to it in the foreseeable future.


> I've written some helper functions for more compact printing of lists and strings. If you are interested, I could put it into a pull request.

Yes please.  I've very interested in seeing pull requests.
